File: src/main/python/ch03/sec03/demo_ngram_segment.py
# -*- encoding: utf-8 -*-    
"""
@Author     :   zYx.Tom
@Contact    :   526614962@qq.com
@site       :   https://zhuyuanxiang.github.io
---------------------------
@Software   :   PyCharm
@Project    :   Hanlp-Books-Examples
@File       :   demo_ngram_segment.py
@Version    :   v0.1
@Time       :   2020-12-22 9:53
@License    :   (C)Copyright 2018-2020, zYx.Tom
@Reference  :   
@Desc       :   
@理解：
"""
import os

from pyhanlp import HanLP

from ch03.sec03.demo_corpus_loader import my_cws_corpus
from preamble import *
from tools import beep_end, test_data_path


# ----------------------------------------------------------------------
def train_bigram(corpus_path, model_path):
    sents = CorpusLoader.convert2SentenceList(corpus_path)
    for sent in sents:
        for word in sent:
            if word.label is None:
                word.setLabel("n")
    marker = NatureDictionaryMaker()
    marker.compute(sents)
    marker.saveTxtTo(model_path)
    pass


def generate_wordnet(sent, trie):
    searcher = trie.getSearcher(JString(sent), 0)
    wordnet = WordNet(sent)
    while searcher.next():
        idx = searcher.begin + 1
        vertex = Vertex(sent[searcher.begin:searcher.begin + searcher.length], searcher.value, searcher.index)
        wordnet.add(idx, vertex)
    # 原子分词，保证图连通
    vertexes = wordnet.getVertexes()
    i = 0
    while i < len(vertexes):
        if len(vertexes[i]) == 0:
            j = i + 1
            for j in range(i + 1, len(vertexes) - 1):  # 寻找第一个非空行 j
                if len(vertexes[j]):
                    break
            vertex = Vertex.newPunctuationInstance(sent[i - 1:j - 1])
            wordnet.add(i, vertex)  # 填充 [i,j) 之间的空白行
            i = j
        else:
            i += len(vertexes[i][-1].realWord)
    return wordnet


def viterbi(wordnet):
    nodes = wordnet.getVertexes()
    # 前向遍历
    for i in range(len(nodes) - 1):
        for node in nodes[i]:
            for to in nodes[i + len(node.realWord)]:
                to.updateFrom(node)  # 根据距离公式计算节点距离，并维护最短路径上的前驱指针from

    # 后向回溯
    path = []  # 最短路径
    f = nodes[len(nodes) - 1].getFirst()  # 从终点回溯
    while f:
        path.insert(0, f)
        f = f.getFrom()  # 按前驱指针 from 回溯
    return [v.realWord for v in path]


def load_bigram(model_path, verbose=True, ret_viterbi=True):
    HanLP.Config.CoreDictionaryPath = model_path + ".txt"
    HanLP.Config.BiGramDictionaryPath = model_path + ".ngram.txt"
    if verbose:
        print(CoreDictionary.getTermFrequency("商品"))
        print(CoreBiGramTableDictionary.getBiFrequency("商品", "和"))
        sent = "商品和服务"
        wordnet = generate_wordnet(sent, CoreDictionary.trie)
        print(wordnet)
        print(viterbi(wordnet))
    viterb_segment = ViterbiSegment().enableAllNamedEntityRecognize(False).enableCustomDictionary(False)
    dijkstra_segment = DijkstraSegment().enableAllNamedEntityRecognize(False).enableCustomDictionary(False)
    return viterb_segment if ret_viterbi else dijkstra_segment


def main():
    corpus_path = my_cws_corpus()
    model_path = os.path.join(test_data_path(), 'my_cws_model')
    train_bigram(corpus_path, model_path)
    load_bigram(model_path)
    pass


if __name__ == '__main__':
    main()
    beep_end()

File: src/main/java/ch03/sec03/DemoNgramSegment.java
package ch03.sec03;

import com.hankcs.hanlp.HanLP;
import com.hankcs.hanlp.corpus.dictionary.NatureDictionaryMaker;
import com.hankcs.hanlp.corpus.document.CorpusLoader;
import com.hankcs.hanlp.corpus.document.sentence.word.IWord;
import com.hankcs.hanlp.dictionary.CoreBiGramTableDictionary;
import com.hankcs.hanlp.dictionary.CoreDictionary;
import com.hankcs.hanlp.seg.Dijkstra.DijkstraSegment;
import com.hankcs.hanlp.seg.Segment;
import com.hankcs.hanlp.seg.Viterbi.ViterbiSegment;

import java.util.List;

import static ch03.MSR.*;
import static tools.Tools.show_subtitle;

public class DemoNgramSegment {

    public static void main(String[] args) {
        // 第一次执行会报出警告，然后会转换 txt 文件为 bin 文件，后面就不再报错
        // 连续执行两次时，第二次使用的是第一次载入的 bin 文件
        // 交换以下现场代码的顺序就可以看到区别
        show_subtitle("my_cws_model");
        trainBigram(MY_CWS_CORPUS_PATH, MY_MODEL_PATH);
        loadBigram(MY_MODEL_PATH);
        // ToDo: HanLP 更新了版本，增加了 CoreDictionary.reload() 函数，我还没有更新版本，没有测试这个功能。
        show_subtitle("msr_ngram");
        trainBigram(MSR_TRAIN_PATH, MSR_MODEL_PATH);
        loadBigram(MSR_MODEL_PATH);
    }

    /**
     * 训练 bigram 模型
     *
     * @param corpusPath 语料库路径
     * @param modelPath  模型保存路径
     */
    public static void trainBigram(String corpusPath, String modelPath) {
        List<List<IWord>> sentenceList = CorpusLoader.convert2SentenceList(corpusPath);
        for (List<IWord> sentence : sentenceList)
            for (IWord word : sentence)
                if (word.getLabel() == null) word.setLabel("n");  // 赋予每个单词一个虚拟的名词词性
        final NatureDictionaryMaker dictionaryMaker = new NatureDictionaryMaker();
        dictionaryMaker.compute(sentenceList);
        dictionaryMaker.saveTxtTo(modelPath);
    }

    public static Segment loadBigram(String modelPath) {
        return loadBigram(modelPath, true, true);
    }

    private static Segment loadBigram(String modelPath, boolean verbose, boolean viterbi) {
//        HanLP.Config.enableDebug();
        HanLP.Config.CoreDictionaryPath = modelPath + ".txt";
        HanLP.Config.BiGramDictionaryPath = modelPath + ".ngram.txt";
        if (verbose) {
            HanLP.Config.ShowTermNature = false;
            System.out.println("「商品」的词频：" + CoreDictionary.getTermFrequency("商品"));
            System.out.println("「商品@和」的频次：" + CoreBiGramTableDictionary.getBiFrequency("商品", "和"));
            Segment segment = new DijkstraSegment()
                    .enableAllNamedEntityRecognize(false)   // 禁用命名实体识别
                    .enableCustomDictionary(false);         // 禁用用户词典
            System.out.println(segment.seg("商品和服务"));
            System.out.println(segment.seg("货币和服务"));
        }
        Segment viterbiSegment = new ViterbiSegment().enableAllNamedEntityRecognize(false).enableCustomDictionary(false);
        Segment dijkstraSegment = new DijkstraSegment().enableAllNamedEntityRecognize(false).enableCustomDictionary(false);
        return viterbi ? viterbiSegment : dijkstraSegment;
    }
}
